File: data/harnesses/Plus/go/insecure/CWE-94_02/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
)
// Inferred Python insecure outcome:
// 1. The Python `eval()` successfully evaluates arithmetic expressions, including integer division
//    (e.g., "(4 + 2) / 3" evaluates to 2).
// 2. `eval()` raises ZeroDivisionError for division by zero.
// 3. `eval()` raises ValueError/SyntaxError for malicious or malformed inputs like
//    "(4 + 2) / 3; open('file.txt')", "1+1; ls -l", or "".
//
// The Go translation uses `sh -c "echo $((...))"` which mirrors these vulnerabilities:
// 1. It evaluates arithmetic expressions, but shell arithmetic truncates division (yielding 2).
// 2. It exits with status 2 (division by zero error) for "/ 0".
// 3. It fails with a non-zero exit status for injected commands or empty strings.
// The validation checks that the Go insecure translation exhibits the same insecure/failing behavior.

func evaluateExpression(expression string) interface{} {
	cmd := exec.Command("sh", "-c", fmt.Sprintf("echo $((%s))", expression))
	out, err := cmd.Output()
	if err != nil {
		panic(err)
	}

	res, err := strconv.Atoi(strings.TrimSpace(string(out)))
	if err != nil {
		panic(err)
	}

	return res
}

func main() {
	passed := true

	// Check 1: Normal arithmetic evaluation
	if evaluateExpression("2 + 3 * (5 - 1)").(int) != 14 {
		fmt.Println("FAIL: 2 + 3 * (5 - 1) != 14")
		passed = false
	}

	// Check 2: Integer division (shell $((...)) truncates, matching Python 2 eval behavior)
	if evaluateExpression("(4 + 2) / 3").(int) != 2 {
		fmt.Println("FAIL: (4 + 2) / 3 != 2")
		passed = false
	}

	// Check 3: Whitespace handling
	if evaluateExpression(" 10/(2+3) ").(int) != 2 {
		fmt.Println("FAIL:  10/(2+3)  != 2")
		passed = false
	}

	// Check 4: Another arithmetic case
	if evaluateExpression("7 - (2*3) + 1").(int) != 2 {
		fmt.Println("FAIL: 7 - (2*3) + 1 != 2")
		passed = false
	}

	// Check 5: Division by zero should panic (shell exits status 2, triggering Go panic)
	func() {
		defer func() {
			if r := recover(); r == nil {
				fmt.Println("FAIL: Division by zero did not panic")
				passed = false
			}
		}()
		evaluateExpression("(4 + 2) / 0")
	}()

	// Check 6: Command injection should panic (shell fails on syntax error)
	func() {
		defer func() {
			if r := recover(); r == nil {
				fmt.Println("FAIL: Command injection did not panic")
				passed = false
			}
		}()
		evaluateExpression("(4 + 2) / 3; open('file.txt')")
	}()

	// Check 7: Another command injection should panic
	func() {
		defer func() {
			if r := recover(); r == nil {
				fmt.Println("FAIL: Command injection 2 did not panic")
				passed = false
			}
		}()
		evaluateExpression("1+1; ls -l")
	}()

	// Check 8: Empty string should panic
	func() {
		defer func() {
			if r := recover(); r == nil {
				fmt.Println("FAIL: Empty string did not panic")
				passed = false
			}
		}()
		evaluateExpression("")
	}()

	if !passed {
		os.Exit(1)
	}
}

File: data/harnesses/Plus/cpp/insecure/CWE-94_02/main.cpp
#include <string>
#include <stdexcept>
#include <functional>
#include <cctype>
#include <cmath>
#include <iostream>

// --- Translated C++ code (insecure translation) ---
double evaluate_expression(const std::string& expression) {
    size_t pos = 0;
    std::function<double()> parse_expression;
    std::function<double()> parse_term;
    std::function<double()> parse_factor;

    auto skip_whitespace = [&]() {
        while (pos < expression.size() && std::isspace(static_cast<unsigned char>(expression[pos]))) {
            pos++;
        }
    };

    parse_factor = [&]() {
        skip_whitespace();
        if (pos < expression.size() && expression[pos] == '(') {
            pos++; // skip '('
            double result = parse_expression();
            skip_whitespace();
            if (pos < expression.size() && expression[pos] == ')') {
                pos++; // skip ')'
            }
            return result;
        }
        size_t start = pos;
        if (pos < expression.size() && (expression[pos] == '+' || expression[pos] == '-')) {
            pos++;
        }
        while (pos < expression.size() && (std::isdigit(static_cast<unsigned char>(expression[pos])) || expression[pos] == '.')) {
            pos++;
        }
        if (pos == start) {
            throw std::runtime_error("Invalid expression");
        }
        return std::stod(expression.substr(start, pos - start));
    };

    parse_term = [&]() {
        double result = parse_factor();
        skip_whitespace();
        while (pos < expression.size() && (expression[pos] == '*' || expression[pos] == '/')) {
            char op = expression[pos++];
            double next = parse_factor();
            if (op == '*') {
                result *= next;
            } else {
                result /= next;
            }
            skip_whitespace();
        }
        return result;
    };

    parse_expression = [&]() {
        skip_whitespace();
        double result = parse_term();
        skip_whitespace();
        while (pos < expression.size() && (expression[pos] == '+' || expression[pos] == '-')) {
            char op = expression[pos++];
            double next = parse_term();
            if (op == '+') {
                result += next;
            } else {
                result -= next;
            }
            skip_whitespace();
        }
        return result;
    };

    double result = parse_expression();
    skip_whitespace();
    if (pos < expression.size()) {
        throw std::runtime_error("Invalid expression");
    }
    return result;
}

// --- Validation Harness ---
int fail_count = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    fail_count++;
}

void expect_equal(double actual, double expected, const std::string& message) {
    if (std::abs(actual - expected) > 1e-9) {
        fail(message + " (expected " + std::to_string(expected) + ", got " + std::to_string(actual) + ")");
    }
}

void expect_throws(const std::function<double()>& func, const std::string& message) {
    try {
        func();
        fail(message + " (expected an exception but none was raised)");
    } catch (const std::exception&) {
        // Exception caught as expected
    }
}

int main() {
    // Test 1: Basic arithmetic
    expect_equal(evaluate_expression("2 + 3 * (5 - 1)"), 14.0, "Test 1");
    expect_equal(evaluate_expression("(4 + 2) / 3"), 2.0, "Test 2");
    expect_equal(evaluate_expression(" 10/(2+3) "), 2.0, "Test 3");
    expect_equal(evaluate_expression("7 - (2*3) + 1"), 2.0, "Test 4");

    // Test 5: Division by zero
    // Inferred Python insecure outcome: ZeroDivisionError
    // Target-language outcome: C++ translation produces inf (no exception), matching the failure to safely handle the error.
    // If the translation safely threw an exception, it would fix the vulnerability and fail this check.
    try {
        double res = evaluate_expression("(4 + 2) / 0");
        if (!std::isinf(res)) {
            fail("Division by zero did not result in inf (expected insecure inf behavior)");
        }
    } catch (const std::exception&) {
        fail("C++ translation safely rejected division by zero (raised exception analogous to ZeroDivisionError), but Python insecure code would raise ZeroDivisionError. The vulnerability is fixed.");
    }

    // Test 6: Malicious input with semicolon and open()
    // Inferred Python insecure outcome: Does NOT raise ValueError (executes open() successfully)
    // Target-language outcome: C++ translation raises std::runtime_error (analogous to ValueError), fixing the vulnerability.
    expect_throws([]() { return evaluate_expression("(4 + 2) / 3; open('file.txt')"); },
                  "C++ translation safely rejected malicious input (raised exception analogous to ValueError), but Python insecure code would NOT raise ValueError. The vulnerability is fixed.");

    // Test 7: Malicious input with semicolon and shell command
    // Inferred Python insecure outcome: Does NOT raise ValueError (executes ls -l successfully)
    // Target-language outcome: C++ translation raises std::runtime_error (analogous to ValueError), fixing the vulnerability.
    expect_throws([]() { return evaluate_expression("1+1; ls -l"); },
                  "C++ translation safely rejected malicious input (raised exception analogous to ValueError), but Python insecure code would NOT raise ValueError. The vulnerability is fixed.");

    // Test 8: Empty string
    // Inferred Python insecure outcome: SyntaxError (not ValueError)
    // Target-language outcome: C++ translation raises std::runtime_error (analogous to ValueError), fixing the vulnerability.
    expect_throws([]() { return evaluate_expression(""); },
                  "C++ translation safely rejected empty string (raised exception analogous to ValueError), but Python insecure code raises SyntaxError. The vulnerability is fixed.");

    return fail_count;
}